src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -679,134 +737,100 @@ createVersionedTable tablename db = do
     indexcreationstmt =
         "CREATE INDEX IF NOT EXISTS " <> tbl ixName <> " ON " <> tbl tablename <> "(txid DESC);"
 
--- | Rewind the checkpoint in the current BlockHistory to the requested block
--- height and hash. This doesn't handle forks. The requested block hash must
--- exist in the history. Otherwise this function will fail.
---
--- /Precondition:/
---
--- For @handlePossibleRewind v c height parentHash@ the respective block for
--- @parentHash@ must be of blockHeight @height - 1@.
---
-handlePossibleRewind
-    :: HasCallStack
-    => ChainwebVersion
-    -> ChainId
-    -> BlockHeight
-        -- ^ The block height to which the check pointer is restored. This is the
-        -- height off the block that is going to be validated.
-
-    -> ParentHash
-        -- ^ The parent of the block that is going to be validated. The height
-        -- of the respective block is one less than the height provided in the
-        -- previous argument.
-
-    -> BlockHandler logger SQLiteEnv TxId
-handlePossibleRewind v cid bRestore hsh = do
-    bCurrent <- getBCurrentHeight
-    checkHistoryInvariant (bCurrent + 1)
-    case compare bRestore (bCurrent + 1) of
-        GT -> internalError "handlePossibleRewind: Block_Restore invariant violation!"
-        EQ -> newChildBlock bCurrent
-        LT -> rewindBlock bRestore
+-- | set up the block state to prepare for a block being played on top of
+-- a ParentContext.
+prepareToPlayBlock
+  :: T.Text
+  -> ChainwebVersion
+  -> ChainId
+  -> Maybe ParentHeader
+  -> BlockHandler logger SQLiteEnv ()
+prepareToPlayBlock msg v cid mph = do
+  let currentHeight = maybe (genesisHeight v cid) (succ . _blockHeight . _parentHeader) mph
+  assign bsBlockHeight currentHeight
+  bsModuleNameFix .= enableModuleNameFix v cid currentHeight
+  bsSortedKeys .= pact42 v cid currentHeight
+  bsLowerCaseTables .= chainweb217Pact v cid currentHeight
+  txid <- getEndTxId (msg <> ".prepareToPlayBlock") mph
+  assign bsTxId txid
+  clearPendingTxState
+
+
+-- | delete any state from the database newer than the input ParentContext.
+rewindDbTo
+    :: Maybe ParentHeader
+    -> BlockHandler logger SQLiteEnv ()
+rewindDbTo Nothing = do
+    withSavepoint DbTransaction $
+      callDb "doRestoreInitial: resetting tables" $ \db -> do
+      exec_ db "DELETE FROM BlockHistory;"
+      exec_ db "DELETE FROM [SYS:KeySets];"
+      exec_ db "DELETE FROM [SYS:Modules];"
+      exec_ db "DELETE FROM [SYS:Namespaces];"
+      exec_ db "DELETE FROM [SYS:Pacts];"
+      tblNames <- qry_ db "SELECT tablename FROM VersionedTableCreation;" [RText]
+      forM_ tblNames $ \t -> case t of
+        [SText tn] -> exec_ db ("DROP TABLE [" <> tn <> "];")
+        _ -> internalError "Something went wrong when resetting tables."
+      exec_ db "DELETE FROM VersionedTableCreation;"
+      exec_ db "DELETE FROM VersionedTableMutation;"
+      exec_ db "DELETE FROM TransactionIndex;"
```

Comment:
We need to assign `bsTxId` and `bsBlockHeight` in this case both to match the previous behavior, but it should also be more robust.

Reply:
I don't have the block height here unfortunately. Let's just rely on `prepareToPlayBlock` setting these and delete the setting here.